tests/testutility.nim

```diff
@@ -71,3 +72,88 @@ suite "Utility":
   test "unsuccessful safeConvert from uint to int":
     check not (compiles do:
       result: uint = safeConvert[int, uint](11.uint))
+
+suite "withValue and valueOr templates":
+  type
+    TestObj = ref object
+      x: int
+
+  proc objIncAndOpt(self: TestObj): Opt[TestObj] =
+    self.x.inc()
+    return Opt.some(self)
+
+  proc objIncAndOption(self: TestObj): Option[TestObj] =
+    self.x.inc()
+    return some(self)
+
+  test "withValue calls right branch when Opt/Option is none":
+    var counter = 0
+    # check Opt/Option withValue with else
+    Opt.none(TestObj).withValue(v):
+      check false
+    else:
+      counter.inc()
+    none(TestObj).withValue(v):
+      check false
+    else:
+      counter.inc()
+    check counter == 2
+
+    # check Opt/Option withValue without else
+    Opt.none(TestObj).withValue(v):
+      check false
+    none(TestObj).withValue(v):
+      check false
+
+  test "withValue calls right branch when Opt/Option is some":
+    var obj = TestObj(x: 0)
```

Comment:
Do you know if we need this object? Could we create an Option of an int and use a counter like you did in the previous test?